src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -376,429 +337,218 @@ public int compare(final byte[] o1, final byte[] o2) {
-            this.length = length;
+            for (final ReadTag tag : record.tags) {
+                if (tag.keyType3BytesAsInt != id) {
+                    continue;
+                }
+                final byte[] data = tag.getValueAsByteArray();
+                try {
+                    os.write(data);
+                } catch (final IOException e) {
+                    throw new RuntimeException(e);
+                }
+                if (!byteSizes.containsKey(data.length)) {
+                    byteSizes.put(data.length, new MutableInt());
+                }
+                byteSizes.get(data.length).value++;
+            }
         }
 
-        @Override
-        public int compareTo(@SuppressWarnings("NullableProblems") final BitCode o) {
-            final int result = value - o.value;
-            if (result != 0)
-                return result;
-            return length - o.length;
-        }
-    }
+        final byte[] data = os.toByteArray();
 
-    public static class HuffmanParamsCalculator {
-        private final HashMap<Integer, MutableInt> countMap = new HashMap<Integer, MutableInt>();
-        private int[] values = new int[]{};
-        private int[] bitLens = new int[]{};
+        final ExternalCompressor gzip = ExternalCompressor.createGZIP();
+        final int gzipLen = gzip.compress(data).length;
 
-        public void add(final int value) {
-            MutableInt counter = countMap.get(value);
-            if (counter == null) {
-                counter = new MutableInt();
-                countMap.put(value, counter);
-            }
-            counter.value++;
-        }
+        final ExternalCompressor rans0 = ExternalCompressor.createRANS(RANS.ORDER.ZERO);
+        final int rans0Len = rans0.compress(data).length;
 
-        public void add(final Integer value, final int inc) {
-            MutableInt counter = countMap.get(value);
-            if (counter == null) {
-                counter = new MutableInt();
-                countMap.put(value, counter);
-            }
-            counter.value += inc;
-        }
+        final ExternalCompressor rans1 = ExternalCompressor.createRANS(RANS.ORDER.ONE);
+        final int rans1Len = rans1.compress(data).length;
 
-        public int[] bitLens() {
-            return bitLens;
-        }
+        final EncodingDetails d = new EncodingDetails();
 
-        public int[] values() {
-            return values;
+        final int minLen = Math.min(gzipLen, Math.min(rans0Len, rans1Len));
+        if (minLen == rans0Len) {
+            d.compressor = rans0;
+        } else if (minLen == rans1Len) {
+            d.compressor = rans1;
+        } else {
+            d.compressor = gzip;
         }
 
-        public Integer[] valuesAsAutoIntegers() {
-            final Integer[] intValues = new Integer[values.length];
-            for (int i = 0; i < intValues.length; i++)
-                intValues[i] = values[i];
+        switch (type) {
+            case 'A':
```

Comment:
It looks like this switch statement might be clearer as a check for membership in a set of these bytes.